Migrates the improved Transaction Generator using Fixed points into master.

TimSheard opened the pull request.

TimSheard commented:
This PR pulls a new way to generate random Transactions. The problem is that when one generates a transaction, one cannot determine the fee until the whole transaction is generated, but the fee is part of the transaction, and depends upon the whole transaction. We apply a fixed point algorithm.  Using a draft fee, and updating until we reach a fixed point.
This approach was first used in the experimental branch fixed-point-gentx.

It was created by coying master, and then cherry picking 5 files from the experimental branch.
   chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/BenchmarkFunctions.hs
   chain-and-ledger/shelley-spec-ledger-test/bench/Shelley/Spec/Ledger/Bench/Gen.hs
   chain-and-ledger/shelley-spec-ledger-test/bench/Main.hs
   chain-and-ledger/shelley-spec-ledger-test/bench/BenchValidation.hs
   chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs

Quite a bit of work had to be applied to get the types to agree, since the experimental branch had a different notion of Era  and the Val class.

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -50,6 +51,7 @@ import Shelley.Spec.Ledger.BaseTypes
   )
 import Shelley.Spec.Ledger.Coin (Coin (..))
 import Shelley.Spec.Ledger.Credential (Credential (..), StakeReference (..))
+-- import Shelley.Spec.Ledger.Crypto (Crypto)
```
we can just remove this

TimSheard replied:
OK done.

JaredCorduan commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -210,61 +175,222 @@ genTxRetry
-              <> (fold (snd <$> wdrls)) Val.~~ deposits
-              <> refunds
+      let spendingBalance = spendingBalanceUtxo  <+> (inject $ ((sumVal (snd <$> wdrls)) <-> deposits) <+> refunds)
+          n = if (Map.size . unUTxO) utxo < 100 then 3 else 0
```
should we maybe make this constants somewhere? add to the test parameter class? or just give them names so that they don't seem so much like magic numbers?

TimSheard replied:
I have added to the defaultConstants
  -- | When generating Tx, we want the UTxO size to fluctuate around this point. If
    --   it gets too small, we can't balance the fee, too large it gets too complicated.
    genTxStableUtxoSize:: Int,
    -- | If we need to grow the Utxo when generating a Tx, how much should it grow by.
    genTxUtxoIncrement:: Int

JaredCorduan reviewed (approved):
I made two very minor comments, but it looks great to me. This is a really great change, thank you!

uroboros commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs:
```diff
@@ -74,16 +78,38 @@ class
+x <+> y = x <> y
+
+(<->) :: Val t => t -> t -> t
+x <-> y = x <+> (scale (-1) y)
```
```suggestion
x <-> y = x <+> (neg y)
```

TimSheard replied:
Yes that is nice.

uroboros commented on shelley/chain-and-ledger/shelley-spec-ledger-test/shelley-spec-ledger-test.cabal:
```diff
@@ -115,6 +117,10 @@ test-suite shelley-spec-ledger-test
       Test.Shelley.Spec.Ledger.Examples.Updates
       Test.Shelley.Spec.Ledger.Fees
       Test.Shelley.Spec.Ledger.MultiSigExamples
+      Test.Shelley.Spec.Ledger.NonTraceProperties.Generator
```
While you were making these changes originally, I went to some trouble to remove the NonTraceProperties (and Hedgehog too). Rather than using these generators you can use one of the trace generators (which generate richer data) and remove NonTraceProperties.*

TimSheard replied:
We can try this in another PR.  Don't want to mix things up too much.

uroboros commented on shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/NonTraceProperties/Generator.hs:
```diff
@@ -95,13 +95,20 @@ import Test.Shelley.Spec.Ledger.Generator.Core
 import Test.Shelley.Spec.Ledger.Utils
 import Unsafe.Coerce
 
+-- ==============================================================
```
(will have to move this elsewhere and remove NonTraceProperties)

TimSheard replied:
In another PR

uroboros reviewed (changes requested):
The fixed point generator looks great! Just need to revert adding of NonTraceProperties

uroboros commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs:
```diff
@@ -19,6 +19,10 @@
 module Cardano.Ledger.Val
   ( Val (..),
     (~~),
+    (<+>),
```
<+> <-> are nice

TimSheard replied:
I also added   (a <×> b) for (scale a b)

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs:
```diff
@@ -74,16 +79,43 @@ class
   size :: t -> Integer -- compute size of Val instance
   -- TODO add PACK/UNPACK stuff to this class
 
+  vsplit :: t -> Integer -> ([t], Coin)
```
This needs documentation. It's not at all clear from the type signature what this is meant to do!

nc6 replied:
Per below comment, just remove it.

TimSheard replied:
So I have moved this to the file 
cardano-ledger-specs/shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs.
It is well commented there.
It becomes a method of a new class Split. There is an instance for Splt for Coin. Adding new instances for Split will help support richer generators for Transaction like things when we add MultiAssets.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs:
```diff
@@ -74,16 +79,43 @@ class
   coin = id
   inject = id
   size _ = 1
+  vsplit (Coin n) 0 = ([], Coin n)
```
I thought splitting was only used in tests? E.g. Polina added it to the test generator in https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs#L112.

The only place this function is used, it could take advantage of the `splitCoin` from that module.

TimSheard replied:
The split coin in that module does something different. I removed it from the class, and added a versionto the only file that uses this.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs:
```diff
@@ -74,16 +79,43 @@ class
+(<×>) :: Val t => Int -> t -> t
+x <×> y = scale x y
+
+neg :: Val t => t -> t
```
This is already defined as `invert`. Do we really need another name? `neg` is only used in one place. Would just remove and use `invert`.

TimSheard replied:
An oversight. We use invert for everything. We export a version typed for Val rather than an aribrary Group instance.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Val.hs:
```diff
@@ -74,16 +79,43 @@ class
+
 -- | Group subtraction. When we move to groups-0.5 we can export this from
 -- there.
 (~~) :: Group g => g -> g -> g
```
If we're adding `(<->)` as the universal `Val` subtraction, then let's just remove `(~~)`. Otherwise this is just confusing

TimSheard replied:
Now there is only <->.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/bench/BenchValidation.hs:
```diff
@@ -1,11 +1,15 @@
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
```
? This is not the formatting we use

TimSheard replied:
Removed

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/bench/Main.hs:
```diff
@@ -196,12 +204,13 @@ epochAt x =
-dstate' :: DState C
-pstate' :: PState C
-utxo' :: UTxO C
+{-
```
No point having dead code in comments, just delete it.

TimSheard replied:
removed

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/NonTraceProperties/Generator.hs:
```diff
@@ -95,13 +95,20 @@ import Test.Shelley.Spec.Ledger.Generator.Core
   ( applyTxBody,
     genesisCoins,
   )
-import Test.Shelley.Spec.Ledger.Generator.Utxo (splitCoin)
```
Is this to resolve an import cycle? I don't see an obvious one

nc6 replied:
Resolved now that @uroboros has removed this module!

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -210,61 +174,224 @@ genTxRetry
+      let fee = _txfee (_body draftTx2)
+          utxoSize = (Map.size . unUTxO) utxo
+          infrequent = hash(fromIntegral (unSlotNo slot) + fromIntegral (unCoin fee) + utxoSize)::Int
+      -- infrequently we trace the final fee and the UTxO size to show progress, and to be sure the UtxO does not become empty.
```
I would say in general we should avoid committing debug traces, even in test code.

TimSheard replied:
The algorithm has an instability in that if we don't balance  genTxStableUtxoSize  and  genTxUtxoIncrement  correctly the size of the UTxO gradually shrinks so small we cannot support generating a transaction. This writes out a bit of information about once in every 10000 calls that lets the observer watch the trend in the growth (or shrinking) of the UTxO.  This is invaluable information.  I can imagine some one changing these, or some other change that upsets this balance. I will comment this out, but not remove it.

JaredCorduan replied:
I think that for a long time we have all been operating under the implicit assumption that we would not check in commented code, even for debugging test code.

TimSheard replied:
I left a comment about what might go wrong. I removed the code that can test for that.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -346,21 +475,22 @@ genTxBody inputs outputs certs wdrls update fee slotWithTTL mdHash = do
 calcOutputsFromBalance ::
-  (HasCallStack, Era era) =>
-  Coin ->
+  (Era era,Val v) =>
```
I guess in theory it's nice that this function is `Val`-polymorphic, but in reality it's not - we project down to coin to construct the `TxOut`s. It's only ever called with `Coin`, and the cost of this change is to introduce regressions elsewhere - moving test code back into the main library, adding things back into `NonTraceProperties` etc. So better to leave it fixed at `Coin` for now, and let it be made appropriate polymorphic at the same time as the rest of the testing code.

nc6 commented on shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:
```diff
@@ -372,22 +502,22 @@ calcOutputsFromBalance balance_ addrs fee =
-genInputs ::
-  (HasCallStack, Era era) =>
-  Constants ->
+genInputs :: forall v era.
```
Again, this type signature is misleading. It claims to be a generator over an arbitrary value `v`, but it's not. The values that it generates are just `Coin`, albeit wrapped up in a `v`-shaped disguise. 

Better for the type signature to be clear about what it's generating. Then when we come to actually make the tests run over multiple eras, the type system guides us to which parts need changing.

TimSheard replied:
I don't agree. This is a case of anticipating possible changes, at no cost.

nc6 reviewed (dismissed):
The meat of this change (the improved generator) looks great, but a number of regressions have slipped through that need to be removed before merging.

uroboros committed:
Added a new improved generator for transactions.

    This algorithm uses a fixed-point strategey, making small changes to the transaction
    fee (which is both included in the transaction, and whose value is based upon the total
    transaction) until a fixed is reached.
    We also added a few operations to the Val class to make this possible. This included
    some infix operators as synonyms which cleaned a lot of code.

uroboros reviewed (approved):
Thanks @TimSheard
